--- node-v10.15.3/node.gyp.old	2019-03-05 15:16:32.000000000 +0000
+++ node-v10.15.3/node.gyp	2019-04-06 13:54:51.204939193 +0000
@@ -479,6 +479,11 @@
       'msvs_disabled_warnings!': [4244],
 
       'conditions': [
+        [ 'host_arch=="mips" or host_arch=="mipsel" or host_arch=="ppc"', {
+	  'link_settings': {
+	    'libraries': [ '-latomic' ],
+	  },
+	}],
         [ 'node_code_cache_path!=""', {
           'sources': [ '<(node_code_cache_path)' ]
         }, {
