diff -Naur a/netsurf/frontends/gtk/download.c b/netsurf/frontends/gtk/download.c
--- a/netsurf/frontends/gtk/download.c	2019-07-17 15:37:00.000000000 +0000
+++ b/netsurf/frontends/gtk/download.c	2019-12-18 15:49:13.389999922 +0000
@@ -751,8 +751,8 @@
 static gboolean nsgtk_download_gsourcefunc__nsgtk_download_update(
 		gpointer user_data)
 {
-	bool *force_update = user_data;
-	return nsgtk_download_update(*force_update);
+	bool force_update = (bool)user_data;
+ 	return nsgtk_download_update(force_update);
 }
 
 static struct gui_download_window *
@@ -838,7 +838,7 @@
 		g_timeout_add(
 			UPDATE_RATE,
 			nsgtk_download_gsourcefunc__nsgtk_download_update,
-			FALSE);
+			(gpointer)FALSE);
 	}
 
 	nsgtk_downloads_list = g_list_prepend(nsgtk_downloads_list, download);
